File: lecture_013/shoePair.java
//  find the missing pair int the array

import java.util.Scanner;

/**
 * shoePair
 */
public class shoePair {
	public static Scanner scn = new Scanner(System.in);

	public static void main(String[] args) {
		int[] arr = {1,1,2,2,3,3,4,4,5,5,7,8,8,9,9,10,10,15,15};
		System.out.println(shoePair(arr));
	}

	public static int shoePair(int[] arr) {
		if (arr.length % 2 == 0) {
			System.out.println("all paired");
			return -1;
		}
		int si = 0, ei = arr.length - 1;
		while (si <= ei) {
			int mid = (si + ei) / 2;
			if (si == ei) {
				return mid;
			}
			else if (mid % 2 == 0 && arr[mid] == arr[mid + 1]) {
				si = mid + 1;
			}
			else if (mid % 2 == 1 && arr[mid] == arr[mid - 1]) {
				si = mid + 1;
			}
			else {
				ei = mid - 1;
			}
		}
		return -1;
	}
}

File: lecture_013/shoePair.cpp
//  find missing pair in the array

#include<iostream>
#include<vector>
#include<string>
using namespace std;

int missingPair(vector<int> &arr) {
	if (!(arr.size() % 2)) {
		cout<<"all paired"<<endl;
		return -1;
	}

	int si = 0, ei = arr.size() - 1;
	while (si <= ei) {
		int mid = (si + ei)/2;
		if (si == ei) {
			return mid;
		}
		/*
		else if (!(mid % 2)) {
			if (arr[mid] == arr[mid + 1]) {
				si = mid + 1;
			}
			else {
				ei = mid - 1;
			}
		}
		else if (mid % 2) {
			if (arr[mid] == arr[mid - 1]) {
				si = mid + 1;
			}
			else {
				ei = mid - 1;
			}
		}
		*/
		//  short form for above commented code
		else if ((!(mid % 2) && arr[mid] == arr[mid + 1]) || ((mid % 2) && arr[mid] == arr[mid - 1])) {
			si = mid + 1;
		}
		else {
			ei = mid - 1;
		}
	}
	return -1;
}

int main(int args, char**argv) {
	vector<int> arr = {1,1,2,2,3,3,4,4,5,5,7,8,8,10,10,11,11};
	cout<<missingPair(arr)<<endl;
	return 0;
}